enhancements/sig-architecture/58-addon-configuration/README.md

```diff
@@ -0,0 +1,589 @@
+this annotation represents a toleration array with json format
+
+Example:
+
+```yaml
+apiVersion: addon.open-cluster-management.io/v1alpha1
+kind: ManagedClusterAddOn
+metadata:
+  name: myaddon
+  namespace: cluster1
+  annotations:
+    addon.open-cluster-management.io/nodeSelector: "{\"kubernetes.io/os\":\"linux\"}"
+    addon.open-cluster-management.io/tolerations: "[{\"key\":\"node.kubernetes.io/unreachable\",\"operator\":\"Exists\",\"effect\":\"NoSchedule\"}]"
+spec:
+  installNamespace: open-cluster-management-agent-addon
+```
+
+In the addon-framework, there is a function [`GetValuesFromAddonAnnotation`](https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/helper.go#L21),
+it returns a [`Values`]([https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/addonfactory.go#L22)
+object (the object type is `map[string]interface{}`) that is built from the add-on values annotation, this object is used
+to render the add-on deployment.
+
+In this alternative, if one add-on has the node selector and tolerations annotations, this function will merge them to
+the `Values` object with the key `NodeSelector` and `Tolerations`.
+
+For an add-on implementation, developers use [`AgentAddonFactory`](https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/addonfactory.go#L28)
+to build their add-on, they use [`WithGetValuesFuncs`](https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/addonfactory.go#L61)
+to specify the get values functions (the function type is [`GetValuesFunc`](https://github.com/open-cluster-management-io/addon-framework/blob/v0.3.0/pkg/addonfactory/addonfactory.go#L24))
+to render their add-on deployment, for get values function, the developers can choose to use addon-frameworks
+`GetValuesFromAddonAnnotation` function or implement it by themselves to handle node selector and toleration for their
+add-ons.
+
+For upgrade, if there is an existing add-on CR that already used the add-on values annotation to specify the node
+selector and tolerations and the add-on implementation used the `GetValuesFromAddonAnnotation` function to get them. The
+add-on user can choose to keep them in add-on values annotation or use new add-on placement annotation to define them
+and remove them from the add-on values annotation.
+
+Pros:
+- We do not need to change current API definitions.
+- Addon developers can choose their add-on configuration implementation freely.
+
+Cons:
+- It's hard to validate the json string that is in the annotations.
+
+### Adding the node selector and toleration to the `ManagedClusterAddOn` spec
+
+As an alternative, we may add the node selector and toleration to the add-on spec, but this may not meet all needs, e.g.
+if one add-on need to deploy more than one agent, and these agents have different node selector/toleration, and the
```

Comment:
If an add-on needs more than one node selector, how will that be achieved with the API proposed here?

Reply:
for this case, the `AddOnDeploymentConfig` is not suitable, user can choose to define their own config api, then refer to the config api in their managedclusteraddon, so there is the reason why we don't want to add the config information to the managedclusteraddon api directly.